qmt/data/geometry.py

```diff
@@ -275,6 +278,123 @@ def write_fcstd(self, file_path=None):
+                    part_polygons[part_name] = polygons
+                    
+        # Let's deal with the physical domains first, which can have cavities
+        # Loop invariance: at the start and end of the loop, the polygons in
+        # part_polygons that don't contain any other polygons should be added to the
+        # 2d geometry
+        build_order = []
+        while part_polygons:
+            # find the polygons that don't contain any other polygons. They can be
+            # constructed as is
+            for name, poly_list in part_polygons.items():
+                for poly in poly_list:
+                    contains_poly = False
+                    for check_poly in chain.from_iterable(part_polygons.values()):
+                        if check_poly is poly:
+                            continue
+                        if poly.contains(check_poly) and not poly.equals(check_poly):
+                            contains_poly = True
+                            break
+                    if contains_poly == False:
+                        polys_2d[name].append(poly)
+
+            # Among these polygons, if there're polygons that are equal to each other,
+            # give priority to the higher build_order, and delete the other one
+            polys_2d_with_name = []
+            for name, poly_list in polys_2d.items():
+                for poly in poly_list:
+                    polys_2d_with_name.append([name, poly])
+            for (name_1, poly_1), (name_2, poly_2) in combinations(polys_2d_with_name, r=2):
+                if poly_1.equals(poly_2):
+                    ind1 = self.build_order.index(name_1)
+                    ind2 = self.build_order.index(name_2)
+                    if ind1 >= ind2:
+                        polys_2d[name_1].remove(poly_1)
+                    else:
+                        polys_2d[name_2].remove(poly_2)
+
+            # We remove all polygons in the union of polygons that didn't contain other
+            # polygons, and add them to polys_2d. This covers cases like this
+            #  ___________        ___________
+            # |  _______  |      |           |
+            # | |___|___| |  =>  |           |  =>  done
+            # |___________|      |___________|
+            rm_polys_union = unary_union(list(chain.from_iterable(polys_2d.values())))
+            done_parts = []
+            for name, poly_list in part_polygons.items():
+                part_polygons[name] = [poly for poly in poly_list if not rm_polys_union.contains(poly)]
+                if not part_polygons[name]:
+                    build_order.append(name)
+                    done_parts.append(name)
+            for name in done_parts:
+                del part_polygons[name]
+        
+        # Now we deal with the virtual parts, which are just added as is
+        for name, poly_list in virtual_part_polygons.items():
+            build_order.append(name)
+            polys_2d[name] = poly_list
+
```

Comment:
I'm not reproducing the boundary conditions/materials from the geo3ddata object since that's not part of the geo2ddata object. We should unify them some time but Andrey and Georg says that geometry is enough for now